libblkid/src/superblocks/bcachefs.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2021 Thomas Weißschuh <thomas@weissschuh.net>
+ *
+ * This file may be redistributed under the terms of the
+ * GNU Lesser General Public License.
+ *
+ * Based on code fragments from bcache-tools by Kent Overstreet:
+ * http://evilpiepirate.org/git/bcache-tools.git
+ * Also based on superblocks/bcache.c.
+ */
+
+#include "superblocks.h"
+
+#include <stddef.h>
+
+#define BCACHEFS_SECTOR_SIZE   512
+#define BCACHEFS_SB_LABEL_SIZE 32
+
+struct bcachefs_sb_field {
+	uint64_t	data[0];
+	uint32_t	u64s;
+	uint32_t	type;
+};
+
+struct bcachefs_sb_member {
+	uint8_t	uuid[16];
+	uint8_t	padding[40];
+};
+
+struct bcachefs_sb_field_members {
+	struct	bcachefs_sb_field field;
+	struct	bcachefs_sb_member members[0];
+};
+
+struct bcachefs_super_block {
+	uint8_t		csum[16];
+	uint16_t	version;
+	uint16_t	version_min;
+	uint16_t	pad[2];
+	uint8_t		magic[16];
+	uint8_t		uuid[16];
+	uint8_t		user_uuid[16];
+	uint8_t		label[BCACHEFS_SB_LABEL_SIZE];
+	uint64_t	offset;
+	uint64_t	seq;
+	uint16_t	block_size;
+	uint8_t		dev_idx;
+	uint8_t		nr_devices;
+	uint32_t	u64s;
+	uint64_t	time_base_lo;
+	uint32_t	time_base_hi;
+	uint32_t	time_precision;
+	uint64_t	flags[8];
+	uint64_t	features[2];
+	uint64_t	compat[2];
+	uint8_t		layout[512];
+	struct bcachefs_sb_field _start[0];
+};
+
+/* magic string */
+#define BCACHEFS_SB_MAGIC     "\xc6\x85\x73\xf6\x4e\x1a\x45\xca\x82\x65\xf5\x7f\x48\xba\x6d\x81"
+/* magic string len */
+#define BCACHEFS_SB_MAGIC_LEN (sizeof(BCACHEFS_SB_MAGIC) - 1)
+/* super block offset */
+#define BCACHEFS_SB_OFF       0x1000
+/* supper block offset in kB */
+#define BCACHEFS_SB_KBOFF     (BCACHEFS_SB_OFF >> 10)
+/* magic string offset within super block */
+#define BCACHEFS_SB_MAGIC_OFF (offsetof(struct bcachefs_super_block, magic))
+/* fields offset within super block */
+#define BCACHEFS_SB_FIELDS_OFF (offsetof(struct bcachefs_super_block, _start))
+/* tag value for members field */
+#define BCACHEFS_SB_FIELD_TYPE_MEMBERS 1
+
+#define BYTES(f) ((le32_to_cpu((f)->u64s) * 8))
+
+static void probe_bcachefs_sb_members(blkid_probe pr, struct bcachefs_sb_field *field,
+				      uint8_t dev_idx, unsigned char *sb_end)
+{
+	struct bcachefs_sb_field_members *members = (struct bcachefs_sb_field_members *) field;
+
+	if ((unsigned char *) &members->members + (sizeof(*members->members) * dev_idx) > sb_end)
+		return;
+
+	blkid_probe_set_uuid_as(pr, members->members[dev_idx].uuid, "UUID_SUB");
+}
+
+static void probe_bcachefs_sb_fields(blkid_probe pr, const struct blkid_idmag *mag,
+				     const struct bcachefs_super_block *bcs)
+{
+	unsigned long sb_size = BCACHEFS_SB_FIELDS_OFF + BYTES(bcs);
+	unsigned char *sb = _blkid_probe_get_sb(pr, mag, sb_size);
```

Comment:
@karelzak 
This is probably not supposed to be actually used by probe code.
But it is a perfect fit for this function, any chance it can be made officially usable?